src/rust/src/tracing_ffi.rs

```diff
@@ -0,0 +1,552 @@
+use libc::c_char;
+use std::ffi::CStr;
+use std::path::Path;
+use std::slice;
+use std::str;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use tracing::{
+    callsite::{Callsite, Identifier},
+    field::{FieldSet, Value},
+    level_enabled,
+    metadata::Kind,
+    span::Entered,
+    subscriber::{Interest, Subscriber},
+    Event, Metadata, Span,
+};
+use tracing_appender::non_blocking::WorkerGuard;
+use tracing_core::Once;
+use tracing_subscriber::{
+    filter::EnvFilter,
+    layer::Layer,
+    reload::{self, Handle},
+};
+
+#[cfg(not(target_os = "windows"))]
+use std::ffi::OsStr;
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::ffi::OsStrExt;
+
+#[cfg(target_os = "windows")]
+use std::ffi::OsString;
+#[cfg(target_os = "windows")]
+use std::os::windows::ffi::OsStringExt;
+
+trait ReloadHandle {
+    fn reload(&self, new_filter: EnvFilter) -> Result<(), reload::Error>;
+}
+
+impl<L, S> ReloadHandle for Handle<L, S>
+where
+    L: From<EnvFilter> + Layer<S> + 'static,
+    S: Subscriber,
+{
+    fn reload(&self, new_filter: EnvFilter) -> Result<(), reload::Error> {
+        self.reload(new_filter)
+    }
+}
+
+pub struct TracingHandle {
+    _file_guard: Option<WorkerGuard>,
+    reload_handle: Box<dyn ReloadHandle>,
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_init(
+    #[cfg(not(target_os = "windows"))] log_path: *const u8,
+    #[cfg(target_os = "windows")] log_path: *const u16,
+    log_path_len: usize,
+    initial_filter: *const c_char,
+    log_timestamps: bool,
+) -> *mut TracingHandle {
+    let initial_filter = unsafe { CStr::from_ptr(initial_filter) }
+        .to_str()
+        .expect("initial filter should be a valid string");
+
+    if log_path.is_null() {
+        tracing_init_stdout(initial_filter, log_timestamps)
+    } else {
+        let log_path = unsafe { slice::from_raw_parts(log_path, log_path_len) };
+
+        #[cfg(not(target_os = "windows"))]
+        let log_path = Path::new(OsStr::from_bytes(log_path));
+
+        #[cfg(target_os = "windows")]
+        let log_path = Path::new(OsString::from_wide(log_path));
+
+        tracing_init_file(log_path, initial_filter, log_timestamps)
+    }
+}
+
+fn tracing_init_stdout(initial_filter: &str, log_timestamps: bool) -> *mut TracingHandle {
+    let builder = tracing_subscriber::fmt()
+        .with_ansi(true)
+        .with_env_filter(initial_filter);
+
+    let reload_handle = if log_timestamps {
+        let builder = builder.with_filter_reloading();
+        let reload_handle = builder.reload_handle();
+        builder.init();
+        Box::new(reload_handle) as Box<dyn ReloadHandle>
+    } else {
+        let builder = builder.without_time().with_filter_reloading();
+        let reload_handle = builder.reload_handle();
+        builder.init();
```

Comment:
`Builder::init` may panic. As I mentioned above, I believe panicking across FFI boundaries is UB. I think it would be more correct to use `try_init`, and report the error in a C-compatible way if the call to `try_init` fails.